lib/spack/spack/database.py

```diff
@@ -695,8 +695,32 @@ def _read_spec_from_dict(self, spec_reader, hash_key, installs, hash=ht.dag_hash
 
         # Build spec from dict first.
         spec = spec_reader.from_node_dict(spec_dict)
+
+        self._validate_version(spec)
+
         return spec
 
+    def _validate_version(self, spec):
+        # Specs installed from a git sha without an associated spack version, that cannot
+        # be resolved anymore because the repo has been removed from its package.py, get
+        # replaced by a literal string version, to avoid errors in comparison.
+        # TODO: Remove this ad-hoc logic in Spack 0.21. It's mostly there so that these
+        # spack-versionless specs don't break basic operations. Since Spack 0.20, the
+        # resolved version is stored in the DB, so this is not needed for newer installs.
```

Comment:
I'm hesitant to remove this for a while because it's not just a feature that'll be disabled -- it'll break operations on existing installs w/o a good workaround.  We've maintained good backward compatibility for this particular scenario for a long time -- spack can read *very* old databases (b/c people keep the installs around).

Reply:
Agreed, that's a good thing. I'll drop the comment.